src/data/reports/glass-data-submission/GLASSDataSubmissionDefaultRepository.ts

```diff
@@ -825,23 +898,143 @@ export class GLASSDataSubmissionDefaultRepository implements GLASSDataSubmission
         });
     }
 
-    private async duplicateProgram(program: ApprovalIds, items: GLASSDataSubmissionItemIdentifier[]) {
-        await promiseMap(items, async item => {
-            const programEvents = await this.getProgramEvents(program.id, item.orgUnit ?? "");
+    private async duplicateTrackerProgram(
+        program: ApprovalIds,
+        items: GLASSDataSubmissionItemIdentifier[]
+    ): Promise<void> {
+        const approvedProgramStages = await this.getProgramStages(program.approvedId);
+        const programStages = await this.getProgramStages(program.id);
+
+        _.forEach(items, async item => {
+            const trackedEntities = await this.getTrackedEntities(program.id, item.orgUnit);
+
+            if (!_.isEmpty(trackedEntities)) {
+                const approvedTrackedEntities = await this.getTrackedEntities(program.approvedId, item.orgUnit);
+
+                !_.isEmpty(approvedTrackedEntities) &&
+                    this.api.post(
+                        "/tracker",
+                        { importStrategy: "DELETE" },
+                        { trackedEntities: approvedTrackedEntities }
+                    );
+
+                const trackedEntitiesToPost: TrackedEntity[] = trackedEntities.map(trackedEntity => ({
+                    ...trackedEntity,
+                    trackedEntity: "",
+                    programOwners: trackedEntity.programOwners.map(programOwner => ({
+                        ...programOwner,
+                        program: program.approvedId,
+                    })),
+                    enrollments: trackedEntity.enrollments.map(enrollment => ({
+                        ...enrollment,
+                        enrollment: "",
+                        trackedEntity: "",
+                        program: program.approvedId,
+                        events: enrollment.events.map(event => {
+                            const programStageName = programStages.find(
```

Comment:
should this be programStageId instead of programStageName. Not a blocker, but would be clearer.

Reply:
oh no. this is the programStageName. i am matching the approved and non approved program stages by name to get the approved program stage as i don't have the approved program stage ids

Reply:
please do not use names. Use id as program stages do not have code you can get the details from these api
https://dev.eyeseetea.com/who-dev-238/api/29/programs/zMD4VltVy3v.json?fields=programStages
https://dev.eyeseetea.com/who-dev-238/api/29/programStages/jp193m62LFY

Reply:
> @deeonwuli  please do not use names. Use id as program stages do not have code you can get the details from these api https://dev.eyeseetea.com/who-dev-238/api/29/programs/zMD4VltVy3v.json?fields=programStages https://dev.eyeseetea.com/who-dev-238/api/29/programStages/jp193m62LFY